src/AESmodes.cpp

```diff
@@ -5,268 +5,811 @@
+        output.clear();
+        return false;
+    }
+    return true;
+}
 
-    // Copy decrypted block to the output
-    for (std::size_t j = 0; j < 16; j++) {
-      outputPadded[j] ^= input[j+((i-1)*16)];
-      output.push_back(outputPadded[j]); 
+/**
+  Cipher with CFB128 mode
+    Guaranteed no exceptions by:
+    handling all exceptions per ERR51-CPP
+        Related: Honoring exception specifications, all exceptions will be caught per ERR55-CPP
+    not throwing exceptions across execution boundaries (library to application) per ERR59-CPP
+    Guaranteeing Strong exception safety per ERR56-CPP
+        Program state will not be modified
+            Input, key, and IV are constant and output vector is cleared when catching an exception
+  @param input: vector of hex values representing plaintext
+  @param output: vector of hex values representing (padded) ciphertext
+  @param key: vector of hex values representing key to use
+  @param IV: initialization vector to use
+  @return True on success
+*/
+bool encrypt_cfb(const std::vector<unsigned char> &input, std::vector<unsigned char> &output,
+                 const std::vector<unsigned char> &key, const std::vector<unsigned char> &IV) noexcept(true) {
+    try {
+
+        // Calculate padding length, then copy input array and padding into plaintext
+        const std::size_t inputSize = input.size();
+        const std::size_t padLength = NUM_BYTES - (inputSize % NUM_BYTES);
+        const std::size_t plaintextLength = inputSize + padLength;
+
+        // Plaintext accomodates both the input and the necessary padding
+        std::vector<unsigned char> plaintext;
+        plaintext.reserve(plaintextLength);
+        plaintext = input;
+
+        // TODO: investigate whether PKCS#7 is the best choice for padding (padding oracle attack)
+        for (std::size_t i = 0; i < padLength; i++) {
+            // PKCS#7 padding (source: https://www.ibm.com/docs/en/zos/2.1.0?topic=rules-pkcs-padding-method)
+            plaintext.push_back(padLength);
+        }
+
+        // Encrypt the first block
+        std::array<unsigned char, NUM_BYTES> block{0};
+        std::array<unsigned char, NUM_BYTES> outputBlock{0};
+
+        std::copy(IV.begin(), IV.end(), block.begin());
+
+        encrypt(block, outputBlock, key);
+
+        for (std::size_t j = 0; j < NUM_BYTES; j++) {
+            // Secure coding: CTR50-CPP. Guarantee that container indices and iterators are within the valid range
+            output.push_back(outputBlock.at(j) ^ plaintext.at(j));
+        }
+
+        // Loop over the number of subsequent blocks
+        for (std::size_t i = 1; i < plaintextLength / NUM_BYTES; i++) {
+            // Loop over block size and fill each block
+            for (std::size_t j = 0; j < NUM_BYTES; j++) {
+                // Secure coding: CTR50-CPP. Guarantee that container indices and iterators are within the valid range
+                block.at(j) = output.at(j + ((i - 1) * NUM_BYTES));
```

Comment:
This looks like 128-bit CFB mode. Could you confirm?

Reply:
Correct